qiskit/primitives/containers/observable.py

```diff
@@ -0,0 +1,164 @@
+        if not isinstance(self._data, Mapping):
+            raise TypeError(f"Observable data type {type(self._data)} is not a Mapping.")
+
+        # If not already set, we compute num_qubits while iterating over data
+        # to avoid having to do another iteration later
+        data_num_qubits = 0
+        for key, value in self._data.items():
+            if not isinstance(key, tuple):
+                raise TypeError("Invalid Observable key type")
+            if len(key) != 2:
+                raise ValueError(f"Invalid Observable key value {key}")
+            # Check tuple pos types after checking length
+            if not isinstance(key[0], tuple) or not isinstance(key[1], str):
+                raise TypeError("Invalid Observable key type")
+            data_num_qubits = max(data_num_qubits, 1 + max(key[0]))
+            if not isinstance(value, Complex):
+                raise TypeError(f"Value {value} is not a complex number")
+        if self._num_qubits is None:
+            self._num_qubits = data_num_qubits
+        elif self._num_qubits < data_num_qubits:
+            raise ValueError("Num qubits is less than the maximum qubit in ObservableTerms")
+
+    @classmethod
+    def coerce(cls, observable: ObservableLike) -> Observable:
+        """Coerce an observable-like object into an :class:`.Observable`.
+
+        Args:
+            observable: The observable-like input.
+
+        Returns:
+            A coerced observables array.
+
+        Raises:
+            TypeError: If the input cannot be formatted because its type is not valid.
+            ValueError: If the input observable is invalid.
+        """
+        return cls._coerce(observable, num_qubits=None)
+
+    @classmethod
+    def _coerce(cls, observable, num_qubits=None):
+        # Pauli-type conversions
+        if isinstance(observable, SparsePauliOp):
+            # TODO: Make sparse by removing identity qubits in keys
+            data = dict(observable.simplify(atol=0).to_list())
+            return cls._coerce(data, num_qubits=observable.num_qubits)
+
+        if isinstance(observable, (Pauli, str, tuple)):
+            return cls._coerce({observable: 1}, num_qubits=num_qubits)
+
+        # Mapping conversion (with possible Pauli keys)
+        if isinstance(observable, Mapping):
+            key_qubits = set()
+            unique = defaultdict(complex)
+            for key, coeff in observable.items():
+                if isinstance(key, Pauli):
+                    # TODO: Make sparse by removing identity qubits in keys
+                    label, phase = key[:].to_label(), key.phase
+                    if phase != 0:
+                        coeff = coeff * (-1j) ** phase
+                    qubits = tuple(range(key.num_qubits))
+                    key = (qubits, label)
+                elif isinstance(key, str):
+                    qubits = tuple(range(len(key)))
+                    key = (qubits, key)
```

Comment:
This should have the same TODO as Pauli.

Reply:
This assumes we always mean "I" in a key to mean identity, which if we do we should be more explicit about (and in general with strings being single characters same length as qubits etc)